File: tests/fixtures/rust/dirty.rs
// Deliberately unsafe/broken fixture — exercises every rust.js finding id.

use std::process::Command;

fn parse(input: &str) -> Result<i32, std::num::ParseIntError> {
    input.parse::<i32>()
}

fn will_panic(input: &str) -> i32 {
    parse(input).unwrap()
}

fn will_also_panic(input: &str) -> i32 {
    parse(input).expect("should parse")
}

unsafe fn read_raw(p: *const i32) -> i32 {
    unsafe { *p }
}

fn lookup_user(id: &str) {
    sqlx::query(&format!("SELECT * FROM t WHERE id = {}", id));
}

fn run_it(user_input: &str) {
    Command::new("sh").arg("-c").arg(user_input).spawn().unwrap();
}

fn insecure_client() {
    let _builder = reqwest::Client::builder()
        .danger_accept_invalid_certs(true);
}

fn make_token() -> u64 {
    let token = rand::random::<u64>();
    token
}

fn debug_print(x: i32) {
    println!("here {}", x);
    dbg!(x);
}

File: tests/fixtures/go/dirty.go
// Deliberately unsafe/broken fixture — exercises every go.js finding id.
//
// Shares a directory (and so a package) with clean.go — see the note there.
package fixtures

import (
	"crypto/md5"
	"crypto/tls"
	"database/sql"
	"fmt"
	"math/rand"
	"net/http"
	"os/exec"
)

func doWork() (int, error) {
	return 1, nil
}

func ignoresError() {
	result, _ := doWork()
	fmt.Println(result)
}

func lookupUser(db *sql.DB, id string) (*sql.Rows, error) {
	return db.Query(fmt.Sprintf("SELECT * FROM t WHERE id = %s", id))
}

func runCommand(userInput string) {
	exec.Command("sh", "-c", userInput).Run()
}

func insecureConfig() *tls.Config {
	return &tls.Config{
		InsecureSkipVerify: true,
	}
}

func hashPassword(password string) [16]byte {
	return md5.Sum([]byte(password))
}

func makeToken() int64 {
	token := rand.Int63()
	return token
}

func mustNotHappen() {
	panic("unreachable")
}

func fetchPage(url string) (*http.Response, error) {
	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

func debugPrint(x int) {
	fmt.Println("here", x)
}

func sendRequest(req *http.Request) (*http.Response, error) {
	return http.DefaultClient.Do(req)
}

func lookupUserTainted(db *sql.DB, id string) (*sql.Rows, error) {
	q := "SELECT * FROM t WHERE id = " + id
	return db.Query(q)
}
